tvx/extract_msg.go

```diff
@@ -130,27 +128,52 @@ func runExtractMsg(c *cli.Context) error {
 				break
 			}
 			if other.From == msg.From && other.Nonce < msg.Nonce {
-				log.Printf("WARN: tipset includes preceding message with lower nonce from same sender; this extraction won't work")
+				included := false
+				for _, o := range neededPrecursorMsgs {
+					if o.Cid() == other.Cid() {
+						included = true
+					}
+				}
+				if !included {
+					neededPrecursorMsgs = append(neededPrecursorMsgs, other)
+				}
 			}
 		}
 		for _, m := range messages.SecpkMessages {
 			if m.Message.Cid() == mid {
 				break
 			}
 			if m.Message.From == msg.From && m.Message.Nonce < msg.Nonce {
-				log.Printf("WARN: tipset includes preceding message with lower nonce from same sender; this extraction won't work")
+				included := false
+				for _, o := range neededPrecursorMsgs {
+					if o.Cid() == m.Message.Cid() {
+						included = true
+					}
+				}
+				if !included {
+					neededPrecursorMsgs = append(neededPrecursorMsgs, &m.Message)
+				}	
 			}
 		}
 	}
 
 	fmt.Println("getting the _before_ filtered state tree")
-	preroot, err := g.GetMaskedStateTree(includedTs.Parents(), retain)
+	tree, err := g.GetStateTreeRootFromTipset(includedTs.Parents())
```

Comment:
* The included tipset is the tipset where the message was included. Let's call this t-0.
* We are walking back one tipset here, t-1.
* GetStateTreeRootFromTipset walks back yet another tipset, t-2.

I think this logic is walking back too many tipsets.